### PR TITLE
Fixed #24866 -- Added Now() database function

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket 24866](https://code.djangoproject.com/ticket/24866)

### CODE REVIEW

File: django/db/models/functions.py [hunk 1]
```diff
@@ -137,3 +137,16 @@ class Upper(Func):
 
     def __init__(self, expression, **extra):
         super(Upper, self).__init__(expression, **extra)
+
+
+class Now(Func):
+    template = 'CURRENT_TIMESTAMP()'
```
Comment (charettes) [inline django/db/models/functions.py]: What about defaulting to `NOW()` instead? This should be compatible with PostgreSQL and MySQL and that's what I would expect a `Now()` function to be doing.

Comment (charettes) [inline django/db/models/functions.py]: This will require an `as_oracle()` with `self.template = 'CURRENT_TIMESTAMP'` on the other end.

Comment (charettes) [inline django/db/models/functions.py]: Actually removing the trailing parentheses (e.g. `CURRENT_TIMESTAMP`) should work on all backends including SQLite.

**Comment by charettes** [comment]
The added tests seem to cover all use cases so I guess this is only missing a release note.

**Comment by adamchainz** [comment]
I put it under "Models", which I guess is right?

File: docs/ref/models/database-functions.txt [hunk 2]
```diff
@@ -117,6 +117,24 @@ Usage example::
...
     margaret smith
 
+Now
+---
+
+.. class:: Now()
+
+.. versionadded:: 1.9
+
+Returns the current date and time. Since this uses your database server's
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: Returns the database server's... (and chop second sentence which is a bit confusing about what "synchronization purposes" menas IMO?)

File: docs/ref/models/database-functions.txt [hunk 3]
```diff
@@ -117,6 +117,24 @@ Usage example::
...
+clock, as opposed to your web server's, it could be more useful for
+synchronization purposes.
+
+Usage example::
+
+    >>> from django.db.models.functions import Now
+    >>> Article.objects.filter(published__lte=Now())
+    [<Article: How to Django>]
+
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: just 1 blank line needed

**Comment by timgraham** [comment]
Looks good, could you squash your commits?

**Comment by akaariai** [comment]
Notably current_timestamp returns start time of transaction on postgres, I guess this isn't true on other backends. On postgres clock_timestamp() can be used to return current time on server.

**Comment by adamchainz** [comment]
Considering the difference on Postgres, have done some in-depth hair-splitting research:

The 2008 (draft) SQL Standard as [linked to by Wikipedia](https://en.wikipedia.org/wiki/SQL#cite_note-43) (it's the only freely available version) states, in Part 2 Section 6.31:

> 1) The <datetime value function>s CURRENT_DATE, CURRENT_TIME, and CURRENT_TIMESTAMP respectively return the current date, current time, and current timestamp...

and

> 3) Let S be an <SQL procedure statement> that is not generally contained in a <triggered action>. All <datetime value function>s that are contained in <value expression>s that are generally contained, without an inter- vening <routine invocation> whose subject routines do not include an SQL function, either in S without an intervening <SQL procedure statement> or in an <SQL procedure statement> contained in the <triggered action> of a trigger activated as a consequence of executing S, are effectively evaluated simultaneously. The time of evaluation of a <datetime value function> during the execution of S and its activated triggers is implementation-dependent.

... i.e. all the calls to `CURRENT_TIMESTAMP` in a *statement* should be expected to be the same, but when *exactly* that is (the start of the statement, when it actually gets called, when it gets parsed, ...) is allowed to vary.

The [Postgres 9.0 docs](http://www.postgresql.org/docs/9.0/static/functions-datetime.html) state that `CURRENT_TIMESTAMP` really means 'the time of the start of the transaction', but there are several extra functions for which level of granularity you actually want to consider as "now". `STATEMENT_TIMESTAMP()` is the equivalent of what the other DBs do.

The [MySQL 5.5 docs](https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html) state that `CURRENT_TIMESTAMP` is an alias for `NOW()`, and that “`NOW()` returns a constant time that indicates the time at which the statement began to execute”.

The [Oracle docs](http://docs.oracle.com/cd/B19306_01/server.102/b14200/functions037.htm) just say “`CURRENT_TIMESTAMP` returns the current date and time in the session time zone” and don't make it clear whether it's at the statement or transaction level. That said, it's probably safe to assume it's at the statement level. I don't have an Oracle server around to check.

The [SQLite docs](https://www.sqlite.org/search?q=current_timestamp) don't mention `CURRENT_TIMESTAMP` outside of DDL statements, but it does work in queries. I tested to check if it's transaction bound, it appears not:

    sqlite> begin;
    sqlite> select current_timestamp;
    2015-05-30 05:19:30
    sqlite> select current_timestamp;
    2015-05-30 05:19:33

I then tested to check if it's statement-bound, and it is (have to go through python since there is no native sleep):

    In [1]: import sqlite3

    In [2]: import time

    In [3]: con = sqlite3.connect(":memory:")

    In [4]: con.create_function("sleep", 1, time.sleep)

    In [5]: c = con.cursor()

    In [6]: c.execute("SELECT CURRENT_TIMESTAMP, SLEEP(1), CURRENT_TIMESTAMP")

    In [7]: c.fetchall()
    Out[7]: [(u'2015-05-30 05:55:51', None, u'2015-05-30 05:55:51')

So, I'm guessing the least surprising thing for Django to do would be to make ``Now()`` mean at the start of the statement, i.e. special case Postgres to ``STATEMENT_TIMESTAMP()``.

**Comment by timgraham** [comment]
Thanks Adam. I think it would be better to handle the question of whether or not to add "TransactionNow" in a separate ticket/pull request. Personally, I don't see it as a common need that needs to be in contrib.postgres.

**Comment by timgraham** [comment]
Could you edit your commit message per our guidelines, "Fixed #24866 -- Added ...."?

**Comment by carljm** [comment]
I find `TransactionNow()` to be more useful than `Now()` in most cases I run into (e.g. it's often useful for two objects created in the same transaction to have the same creation timestamp). In other words, as usual Postgres gets it right and the other databases get it wrong :-)

That said, `TransactionNow()` (like any other SQL function) is now trivial to implement yourself, so I won't be bothered if it doesn't make it into `contrib.postgres`. But on balance I'd be in favor.

**Comment by carljm** [comment]
(The main benefit I'd see to including it is just that its presence in the docs may alert some developers that they even have that option, which they may not realize otherwise.)

**Comment by adamchainz** [comment]
Yes I was thinking along the same lines - firstly, there may be developers experienced with postgres who might consider the cross-compatible design of `Now()` to be wrong and upset that Django doesn't do it "the postgres way", and secondly, by adding and documenting it, it makes discovery easier for developers who are not so experienced with postgres. Also, I'm a bit of a completionist. I will separate it into a second ticket + PR though so it can be evaluated separately.

File: django/db/models/functions.py [hunk 4]
```diff
@@ -137,3 +137,16 @@ class Upper(Func):
...
+class Now(Func):
+    template = 'CURRENT_TIMESTAMP'
+
+    def __init__(self, output_field=None, **extra):
+        if output_field is None:
+            output_field = DateTimeField()
+        super(Now, self).__init__(output_field=output_field, **extra)
+
+    def as_postgresql(self, compiler, connection):
+        self.template = 'STATEMENT_TIMESTAMP()'
```
Comment (timgraham) [inline django/db/models/functions.py]: Could you add a comment with your explanation about why CURRENT_TIMESTAMP is inappropriate on PostgreSQL.

**Comment by timgraham** [comment]
Needs a rebase too, otherwise LGTM.

File: django/db/models/functions.py (hunk 4 above)
Comment (timgraham) [inline django/db/models/functions.py]: Did you miss this comment again?

Comment (adamchainz) [inline django/db/models/functions.py]: Indeed, thanks for the prod!

**Comment by timgraham** [comment]
merged in 23048d186ce0041654a9f547fe3e7177efce3076, thanks!

**Pull request closed without merging**